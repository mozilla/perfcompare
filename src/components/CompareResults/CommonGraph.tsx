import { Chart as ChartJS, LineElement, LinearScale } from 'chart.js';
import 'chart.js/auto';
import * as kde from 'fast-kde';
import { Line } from 'react-chartjs-2';
import { style } from 'typestyle';

import { Spacing } from '../../styles';

ChartJS.register(LinearScale, LineElement);

const styles = {
  container: style({
    display: 'flex',
    marginBottom: Spacing.Medium,
    width: '100%',
  }),
};

function CommonGraph(props: CommonGraphProps) {
  const { baseRevisionRuns, newRevisionRuns } = props;

  const options = {
    plugins: {
      legend: {
        align: 'start' as const,
        position: 'bottom' as const,
      },
      title: {
        align: 'start' as const,
        display: true,
        text: 'Runs Density Distribution',
      },
    },
    responsive: true,
    scales: {
      x: {
        grid: {
          display: false,
          offset: false,
        },
        type: 'linear' as const,
        title: {
          align: 'end' as const,
          display: true,
          text: 'Run Result' as const,
        },
      },
      y: {
        beginAtZero: true,
        grid: {
          drawBorder: false,
          display: false,
          offset: false,
        },
        ticks: {
          beginAtZero: true,
          display: true,
        },
        title: {
          align: 'end' as const,
          display: true,
          text: 'Run Density' as const,
        },
      },
    },
  };

  //////////////////// START FAST KDE ////////////////////////
  const baseRunsDensity = Array.from(
    kde.density1d(baseRevisionRuns.values, {
      bandwidth: baseRevisionRuns.stddev,
      extent: [
        Math.min(...baseRevisionRuns.values),
        Math.max(...baseRevisionRuns.values),
      ],
    }),
  );
  const newRunsDensity = Array.from(
    kde.density1d(newRevisionRuns.values, {
      bandwidth: newRevisionRuns.stddev,
      extent: [
        Math.min(...newRevisionRuns.values),
        Math.max(...newRevisionRuns.values),
      ],
    }),
  );

  //////////////////// END FAST KDE   ////////////////////////

  const data = {
    datasets: [
      {
        label: 'Base',
        data: baseRunsDensity,
        fill: false,
        borderColor: 'rgba(144, 89, 255, 1)',
      },
      {
        label: 'New',
        data: newRunsDensity,
        fill: false,
        borderColor: 'rgba(0, 135, 135, 1)',
      },
    ],
  };

  return (
    <div className={styles.container}>
      <Line options={options} data={data} />
    </div>
  );
}

interface CommonGraphProps {
  baseRevisionRuns: {
    name: string;
    median: number;
    values: number[];
    stddev: number;
    stddevPercent: number;
  };
  newRevisionRuns: {
    name: string;
    median: number;
    values: number[];
    stddev: number;
    stddevPercent: number;
  };
}

export default CommonGraph;
