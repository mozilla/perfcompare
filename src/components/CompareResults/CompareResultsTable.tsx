import Alert from '@mui/material/Alert';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import Grid from '@mui/material/Grid';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableContainer from '@mui/material/TableContainer';

import { RootState } from '../../common/store';
import { useAppSelector } from '../../hooks/app';
import type { CompareResultsState } from '../../types/state';
import CompareTableStatus from './CompareTableStatus';
import PaginatedCompareResults from './PaginatedCompareResults';

function CompareResultsTable(props: CompareResultsProps) {
  const { theme } = props;
  const compareResults: CompareResultsState = useAppSelector(
    (state: RootState) => state.compareResults,
  );

  return (
    <>
      {!compareResults.loading && (
        <>
          <CompareTableStatus />
          <TableContainer component={Paper}>
            <Table
              sx={{ minWidth: 650 }}
              size="small"
              aria-label="a dense table"
            >
              <colgroup>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'20%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'12%' }}/>
      <col style={{ width:'8%' }}/>
   </colgroup>
              <PaginatedCompareResults theme={theme} />
            </Table>
          </TableContainer>
        </>
      )}
      {compareResults.loading && (
        <Box display="flex" justifyContent="center">
          <Grid item xs={1} className="compare-results-error">
            <CircularProgress />
          </Grid>
        </Box>
      )}
      {compareResults.error && (
        <Box display="flex" justifyContent="center">
          <Grid item xs={6} className="compare-results-error">
            <Alert severity="error" sx={{ textAlign: 'center' }}>
              Error: {compareResults.error}
            </Alert>
          </Grid>
        </Box>
      )}
    </>
  );
}

interface CompareResultsProps {
  theme: string;
}

export default CompareResultsTable;
