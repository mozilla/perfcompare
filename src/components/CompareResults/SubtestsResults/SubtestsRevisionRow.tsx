import { useState } from 'react';

import ExpandLessIcon from '@mui/icons-material/ExpandLess';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ThumbDownIcon from '@mui/icons-material/ThumbDown';
import ThumbUpIcon from '@mui/icons-material/ThumbUp';
import TimelineIcon from '@mui/icons-material/Timeline';
import { IconButton, Box } from '@mui/material';
import { style } from 'typestyle';

import { useAppSelector } from '../../../hooks/app';
import { Strings } from '../../../resources/Strings';
import { Colors, Spacing, ExpandableRowStyles } from '../../../styles';
import type { CompareResultsItem } from '../../../types/state';
import RevisionRowExpandable from '.././RevisionRowExpandable';

const revisionsRow = {
  borderRadius: '4px 0px 0px 4px',
  display: 'grid',
  margin: `${Spacing.Small}px 0px`,
};

const typography = style({
  fontFamily: 'SF Pro',
  fontStyle: 'normal',
  fontWeight: 400,
  fontSize: '13px',
  lineHeight: '16px',
});

function getStyles(themeMode: string) {
  const mainBackgroundColor =
    themeMode === 'light' ? Colors.Background200 : Colors.Background200Dark;
  const backgroundColorExpandButton =
    themeMode === 'light' ? Colors.Background300 : Colors.Background100Dark;
  return {
    revisionRow: style({
      ...revisionsRow,
      $nest: {
        '.base-value': {
          backgroundColor: mainBackgroundColor,
        },
        '.cell': {
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
        },
        '.confidence': {
          backgroundColor: mainBackgroundColor,
        },
        '.comparison-sign': {
          backgroundColor: mainBackgroundColor,
        },
        '.delta': {
          backgroundColor: mainBackgroundColor,
        },
        '.expand-button-container': {
          justifyContent: 'right',
        },
        '.new-value': {
          backgroundColor: mainBackgroundColor,
        },
        '.subtests': {
          backgroundColor: mainBackgroundColor,
          borderRadius: '4px 0 0 4px',
          paddingLeft: Spacing.xLarge,
          justifyContent: 'left',
        },
        '.subtests-container': {
          alignItems: 'flex-end',
          backgroundColor: mainBackgroundColor,
          display: 'flex',
        },
        '.status': {
          backgroundColor: mainBackgroundColor,
          justifyContent: 'center',
        },
        '.total-runs': {
          backgroundColor: mainBackgroundColor,
          gap: '8px',
        },
        '.row-buttons': {
          backgroundColor: mainBackgroundColor,
          borderRadius: '0px 4px 4px 0px',
          display: 'flex',
          justifyContent: 'flex-end',
          $nest: {
            '.download': {
              cursor: 'not-allowed',
            },
          },
        },
        '.expand-button': {
          backgroundColor: backgroundColorExpandButton,
        },
        '.status-hint': {
          display: 'inline-flex',
          gap: '6px',
          borderRadius: '4px',
          padding: '4px 10px',
        },
        '.status-hint-improvement': {
          backgroundColor: '#D8EEDC',
        },
        '.status-hint-regression': {
          backgroundColor: '#FFE8E8',
        },
        '.status-hint .MuiSvgIcon-root': {
          height: '16px',
        },
        '.status-hint-improvement .MuiSvgIcon-root': {
          color: '#017A40',
        },
        '.status-hint-regression .MuiSvgIcon-root': {
          // We need to move the icon a bit lower so that it _looks_ centered.
          marginTop: '2px',
          color: '#D7264C',
        },
      },
    }),
    typography: typography,
  };
}

const styles = {
  light: getStyles('light'),
  dark: getStyles('dark'),
};

const stylesCard = ExpandableRowStyles();

function determineStatus(improvement: boolean, regression: boolean) {
  if (improvement) return 'Improvement';
  if (regression) return 'Regression';
  return '-';
}

function determineStatusHintClass(improvement: boolean, regression: boolean) {
  if (improvement) return 'status-hint-improvement';
  if (regression) return 'status-hint-regression';
  return '';
}

function determineSign(baseMedianValue: number, newMedianValue: number) {
  if (baseMedianValue > newMedianValue) return '>';
  if (baseMedianValue < newMedianValue) return '<';
  return '';
}

function SubtestsRevisionRow(props: RevisionRowProps) {
  const { result, gridTemplateColumns } = props;
  const {
    test,
    base_median_value: baseMedianValue,
    base_measurement_unit: baseUnit,
    new_median_value: newMedianValue,
    new_measurement_unit: newUnit,
    is_improvement: improvement,
    is_regression: regression,
    delta_percentage: deltaPercent,
    confidence_text: confidenceText,
    base_runs: baseRuns,
    new_runs: newRuns,
    graphs_link: graphLink,
  } = result;

  const [expanded, setExpanded] = useState(false);

  const toggleIsExpanded = () => {
    setExpanded(!expanded);
  };

  const themeMode = useAppSelector((state) => state.theme.mode);

  return (
    <>
      <Box
        className={`revisionRow ${styles[themeMode].revisionRow} ${styles[themeMode].typography}`}
        sx={{ gridTemplateColumns }}
        role='row'
      >
        <div className='subtests cell' role='cell'>
          <div className='subtests-container'>{test}</div>
        </div>
        <div className='base-value cell' role='cell'>
          {' '}
          {baseMedianValue} {baseUnit}{' '}
        </div>
        <div className='comparison-sign cell' role='cell'>
          {determineSign(baseMedianValue, newMedianValue)}
        </div>
        <div className='new-value cell' role='cell'>
          {' '}
          {newMedianValue} {newUnit}
        </div>
        <div className='status cell' role='cell'>
          <span
            className={`status-hint ${determineStatusHintClass(
              improvement,
              regression,
            )}`}
          >
            {improvement ? <ThumbUpIcon /> : null}
            {regression ? <ThumbDownIcon /> : null}
            {determineStatus(improvement, regression)}
          </span>
        </div>
        <div className='delta cell' role='cell'>
          {' '}
          {deltaPercent} %{' '}
        </div>
        <div className='confidence cell' role='cell'>
          {' '}
          {confidenceText}{' '}
        </div>
        <div className='total-runs cell' role='cell'>
          <span>
            <span title='Base runs'>B:</span>
            <strong>{baseRuns.length}</strong>
          </span>
          <span>
            <span title='New runs'>N:</span>
            <strong>{newRuns.length}</strong>
          </span>
        </div>
        <div className='row-buttons cell'>
          <div className='graph' role='cell'>
            <div className='graph-link-button-container'>
              <IconButton
                title={Strings.components.revisionRow.title.graphLink}
                color='primary'
                size='small'
                href={graphLink}
                target='_blank'
              >
                <TimelineIcon />
              </IconButton>
            </div>
          </div>
        </div>
        <div className='expand-button cell' role='cell'>
          <div
            className='expand-button-container'
            onClick={toggleIsExpanded}
            data-testid='expand-subtests-row-button'
          >
            <IconButton
              title={
                expanded
                  ? Strings.components.expandableRow.title.shrink
                  : Strings.components.expandableRow.title.expand
              }
              color='primary'
              size='small'
            >
              {expanded ? <ExpandLessIcon /> : <ExpandMoreIcon />}
            </IconButton>
          </div>
        </div>
      </Box>
      {expanded && (
        <div
          className={`content-row ${stylesCard.container}`}
          data-testid='expanded-row-content'
        >
          <RevisionRowExpandable result={result} />
        </div>
      )}
    </>
  );
}

interface RevisionRowProps {
  result: CompareResultsItem;
  gridTemplateColumns: string;
}

export default SubtestsRevisionRow;
