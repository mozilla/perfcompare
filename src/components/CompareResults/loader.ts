import { defer } from 'react-router-dom';

import { repoMap, frameworks } from '../../common/constants';
import {
  fetchCompareResults,
  fetchFakeCompareResults,
  fetchRecentRevisions,
} from '../../logic/treeherder';
import { Changeset, Repository } from '../../types/state';
import { FakeCommitHash, Framework } from '../../types/types';

// This function checks and sanitizes the input values, then returns values that
// we can then use in the rest of the application.
function checkValues({
  baseRev,
  baseRepo,
  newRevs,
  newRepos,
  framework,
}: {
  baseRev: string | null;
  baseRepo: Repository['name'] | null;
  newRevs: string[];
  newRepos: Repository['name'][];
  framework: string | number | null;
}): {
  baseRev: string;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRepos: Repository['name'][];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
} {
  if (baseRev === null) {
    throw new Error('The parameter baseRev is missing.');
  }

  if (baseRepo === null) {
    throw new Error('The parameter baseRepo is missing.');
  }

  const validRepoValues = Object.values(repoMap);
  if (!validRepoValues.includes(baseRepo)) {
    throw new Error(
      `The parameter baseRepo "${baseRepo}" should be one of ${validRepoValues.join(
        ', ',
      )}.`,
    );
  }

  if (framework === null) {
    framework = 1; // default to talos so that manually typing the URL is easier
  }

  const frameworkId = +framework as Framework['id'];
  if (Number.isNaN(frameworkId)) {
    throw new Error(
      `The parameter framework should be a number, but it is "${framework}".`,
    );
  }
  const frameworkName = frameworks.find(
    (entry) => entry.id === frameworkId,
  )?.name;

  if (!frameworkName) {
    throw new Error(
      `The parameter framework isn't a valid value: "${framework}".`,
    );
  }
  if (!newRevs.length) {
    return {
      baseRev,
      baseRepo,
      newRevs: [baseRev],
      newRepos: [baseRepo],
      frameworkId,
      frameworkName,
    };
  }

  if (newRevs.length !== newRepos.length) {
    throw new Error(
      'There should be as many "newRepo" parameters as there are "newRev" parameters.',
    );
  }
  if (!newRepos.every((newRepo) => validRepoValues.includes(newRepo))) {
    throw new Error(
      `Every parameter newRepo "${newRepos.join(
        '", "',
      )}" should be one of ${validRepoValues.join(', ')}.`,
    );
  }

  return {
    baseRev,
    baseRepo,
    newRevs,
    newRepos,
    frameworkId,
    frameworkName,
  };
}

// This is essentially a glue to call the related function from
// /logic/treeherder.ts for all the revs we need results for.
async function fetchCompareResultsOnTreeherder({
  baseRev,
  baseRepo,
  newRevs,
  newRepos,
  framework,
}: {
  baseRev: string;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRepos: Repository['name'][];
  framework: Framework['id'];
}) {
  const promises = newRevs.map((newRev, i) =>
    fetchCompareResults({
      baseRev,
      baseRepo,
      newRev,
      newRepo: newRepos[i],
      framework,
    }),
  );
  return Promise.all(promises);
}

const fakeCommitHashes: FakeCommitHash[] = [
  'bb6a5e451dace3b9c7be42d24c9272738d73e6db',
  '9d50665254899d8431813bdc04178e6006ce6d59',
  'a998c42399a8fcea623690bf65bef49de20535b4',
];

// This is essentially a glue to call the related function from
// /logic/treeherder.ts for all the revs we need results for.
async function fetchAllFakeCompareResults() {
  const promises = fakeCommitHashes.map((hash) =>
    fetchFakeCompareResults(hash),
  );
  return Promise.all(promises);
}

// This function is responsible for fetching the data from the URL. It's called
// by React Router DOM when the compare-results path is requested.
// It uses the URL parameters as inputs, and returns all the fetched data to the
// React components through React Router's useLoaderData hook.
export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);
  const useFakeData = url.searchParams.has('fakedata');
  if (useFakeData) {
    const results = await fetchAllFakeCompareResults();
    // They're all based on the same rev
    const baseRev = results[0][0].base_rev;
    // And the same repository
    const baseRepo = results[0][0].base_repository_name;

    const newRevs = fakeCommitHashes;
    const newRepos = results.map((result) => result[0].new_repository_name);
    const frameworkId = results[0][0].framework_id;
    const frameworkName =
      frameworks.find((entry) => entry.id === frameworkId)?.name ?? '';

    return {
      results,
      baseRev,
      baseRepo,
      newRevs,
      newRepos,
      frameworkId,
      frameworkName,
    };
  }

  const baseRevFromUrl = url.searchParams.get('baseRev');
  const baseRepoFromUrl = url.searchParams.get('baseRepo') as
    | Repository['name']
    | null;
  const newRevsFromUrl = url.searchParams.getAll('newRev');
  const newReposFromUrl = url.searchParams.getAll(
    'newRepo',
  ) as Repository['name'][];
  const frameworkFromUrl = url.searchParams.get('framework');

  const { baseRev, baseRepo, newRevs, newRepos, frameworkId, frameworkName } =
    checkValues({
      baseRev: baseRevFromUrl,
      baseRepo: baseRepoFromUrl,
      newRevs: newRevsFromUrl,
      newRepos: newReposFromUrl,
      framework: frameworkFromUrl,
    });

  const resultsPromise = fetchCompareResultsOnTreeherder({
    baseRev,
    baseRepo,
    newRevs,
    newRepos,
    framework: frameworkId,
  });

  // For each of these requests, we get a list of 1 item because we request one
  // specific hash.
  // TODO what happens if there's no result?
  const baseRevInfoPromise = fetchRecentRevisions({
    repository: baseRepo,
    hash: baseRev,
  }).then((listOfRevisions) => listOfRevisions[0]);
  const newRevsInfoPromises = newRevs.map((newRev, i) =>
    fetchRecentRevisions({ repository: newRepos[i], hash: newRev }).then(
      (listOfRevisions) => listOfRevisions[0],
    ),
  );

  const [baseRevInfo, ...newRevsInfo] = await Promise.all([
    baseRevInfoPromise,
    ...newRevsInfoPromises,
  ]);

  return defer({
    results: resultsPromise,
    baseRev,
    baseRevInfo,
    baseRepo,
    newRevs,
    newRevsInfo,
    newRepos,
    frameworkId,
    frameworkName,
  });
}

type DeferredLoaderData = {
  results: Promise<unknown>;
  baseRev: string;
  baseRevInfo: Changeset;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRevsInfo: Changeset[];
  newRepos: Repository['name'][];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
};

//had to be more explicit with the type because the defer
//function returns a an inaccessible type
export type LoaderReturnValue = DeferredLoaderData;
