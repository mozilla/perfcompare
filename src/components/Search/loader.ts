import { repoMap, frameworks } from '../../common/constants';
import { memoizedFetchRevisionForRepository } from '../../logic/treeherder';
import { Changeset, Repository } from '../../types/state';
import { Framework } from '../../types/types';

const DEFAULT_VALUES = {
  newRev: null,
  newRevInfo: null,
  newRepo: 'try' as Repository['name'],
  frameworkId: 1 as Framework['id'],
  frameworkName: 'talos' as Framework['name'],
};

// This function checks and sanitizes the input values, then returns values that
// we can then use in the rest of the application.
function checkValues({
  newRev,
  newRepo,
  frameworkName,
}: {
  newRev: string | null;
  newRepo: Repository['name'] | null;
  frameworkName: Framework['name'] | null;
}): null | {
  newRev: string;
  newRepo: Repository['name'];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
} {
  if (newRev === null || newRepo === null) {
    return null;
  }

  const validRepoValues = Object.values(repoMap);
  if (!validRepoValues.includes(newRepo)) {
    console.warn(`The repository ${newRepo} wasn't found in our list.`);
    return null;
  }

  if (frameworkName === null) {
    frameworkName = DEFAULT_VALUES.frameworkName;
  }

  let frameworkId = frameworks.find(
    (entry) => entry.name === frameworkName,
  )?.id;

  if (frameworkId === undefined) {
    // Default to talos if the entry wasn't found.
    console.warn(
      `The framework entry for ${frameworkName} wasn't found, defaulting to talos.`,
    );
    frameworkId = DEFAULT_VALUES.frameworkId;
    frameworkName = DEFAULT_VALUES.frameworkName;
  }

  return {
    newRev,
    newRepo,
    frameworkId,
    frameworkName,
  };
}

// This function is responsible for fetching the data from the URL. It's called
// by React Router DOM when the compare-results path is requested.
// It uses the URL parameters as inputs, and returns all the fetched data to the
// React components through React Router's useLoaderData hook.
export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);
  const newRevFromUrl = url.searchParams.get('newRev');
  const newRepoFromUrl = url.searchParams.get('newRepo') as
    | Repository['name']
    | null;
  const frameworkFromUrl = url.searchParams.get('frameworkName') as
    | Framework['name']
    | null;

  const checkedValues = checkValues({
    newRev: newRevFromUrl,
    newRepo: newRepoFromUrl,
    frameworkName: frameworkFromUrl,
  });
  if (!checkedValues) {
    return DEFAULT_VALUES;
  }

  const { newRev, newRepo, frameworkId, frameworkName } = checkedValues;

  const newRevInfo = await memoizedFetchRevisionForRepository({
    repository: newRepo,
    hash: newRev,
  });

  if (!newRevInfo) {
    // The search returned no result.
    return DEFAULT_VALUES;
  }

  return {
    newRev,
    newRevInfo,
    newRepo,
    frameworkId,
    frameworkName,
  };
}

// Be explicit with the returned type to control it better than if we were
// inferring it.
export type LoaderReturnValue = {
  newRev: string;
  newRevInfo: Changeset;
  newRepo: Repository['name'];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
};
