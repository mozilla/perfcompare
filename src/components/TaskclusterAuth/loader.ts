import {
  storeUserToken,
  storeUserCredentials,
} from '../../logic/credentials-storage';
import {
  retrieveTaskclusterUserCredentials,
  retrieveTaskclusterToken,
} from '../../logic/taskcluster';

export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);

  const taskclusterCode = url.searchParams.get('code') as string;
  const state = url.searchParams.get('state');

  const requestState = sessionStorage.getItem('requestState');

  const rootUrl = sessionStorage.getItem('taskclusterUrl') as string;

  if (!taskclusterCode) {
    throw new Error(`Error when getting Taskcluster code from URL.`);
  }

  if (!requestState || !state || requestState !== state) {
    throw new Error(
      `Error missing or not matching request state with URL state`,
    );
  }

  const tokenResponse = await retrieveTaskclusterToken(
    rootUrl,
    taskclusterCode,
  );

  storeUserToken(rootUrl, tokenResponse);

  // fetch access token with token Bearer
  const userCredentials = await retrieveTaskclusterUserCredentials(
    rootUrl,
    tokenResponse.access_token,
  );

  storeUserCredentials(rootUrl, userCredentials);

  return userCredentials;
}

export type LoaderReturnValue = Awaited<ReturnType<typeof loader>>;
