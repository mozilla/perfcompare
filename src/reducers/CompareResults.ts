import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { fetchCompareResults } from '../thunks/compareResultsThunk';
import type { CompareResultsState, ResultsHashmap } from '../types/state';

const initialState: CompareResultsState = {
  data: {},
  loading: false,
  error: undefined,
};

const compareResults = createSlice({
  name: 'compareResults',
  initialState,
  reducers: {
    setCompareData(
      state,
      action: PayloadAction<{
        data: ResultsHashmap;
      }>,
    ) {
      state.data = action.payload.data;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchCompareResults.fulfilled, (state, action) => {
        const revisionHash = action.payload[0].new_rev;
        state.data[revisionHash] = action.payload;
        state.loading = initialState.loading;
      })
      .addCase(fetchCompareResults.pending, (state) => {
        state.loading = true;
        state.error = initialState.error;
      })
      .addCase(fetchCompareResults.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = initialState.loading;
      });
  },
});

export const { setCompareData } = compareResults.actions;
export default compareResults.reducer;
