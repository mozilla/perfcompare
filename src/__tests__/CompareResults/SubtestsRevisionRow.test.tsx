import { ReactElement } from 'react';

import { loader } from '../../components/CompareResults/loader';
import SubtestsRevisionRow from '../../components/CompareResults/SubtestsResults/SubtestsRevisionRow';
import getTestData from '../utils/fixtures';
import {
  screen,
  renderWithRouter,
  FetchMockSandbox,
} from '../utils/test-utils';

function renderWithRoute(component: ReactElement) {
  (window.fetch as FetchMockSandbox)
    .get('begin:https://treeherder.mozilla.org/api/perfcompare/results/', [])
    .get('begin:https://treeherder.mozilla.org/api/project/', {
      results: [],
    });

  return renderWithRouter(component, {
    route: '/subtests-compare-results/',
    search: '?baseRev=spam&baseRepo=try&framework=1',
    loader,
  });
}

describe('SubtestsRevisionRow Component', () => {
  it('renders the component with correct data', async () => {
    const { subtestsResult } = getTestData();
    const mockGridTemplateColumns = '1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr';
    renderWithRoute(
      <SubtestsRevisionRow
        result={subtestsResult[0]}
        gridTemplateColumns={mockGridTemplateColumns}
      />,
    );

    // Check if all elements render correctly
    await screen.findByText('dhtml.html');
    expect(await screen.findByText('971.38 ms')).toBeInTheDocument();
    expect(await screen.findByText('982.41 ms')).toBeInTheDocument();
    expect(await screen.findByText('1.14 %')).toBeInTheDocument();
    expect(await screen.findByText('Low')).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it('renders the Regression status with bg color', async () => {
    const { subtestsResult } = getTestData();
    const mockGridTemplateColumns = '1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr';
    renderWithRoute(
      <SubtestsRevisionRow
        result={subtestsResult[2]}
        gridTemplateColumns={mockGridTemplateColumns}
      />,
    );

    // Check if all elements render correctly
    await screen.findByText('regression.html');
    expect(await screen.findByText('Regression')).toBeInTheDocument();
  });

  it('renders the Improvement status with bg color', async () => {
    const { subtestsResult } = getTestData();
    const mockGridTemplateColumns = '1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr';
    renderWithRoute(
      <SubtestsRevisionRow
        result={subtestsResult[3]}
        gridTemplateColumns={mockGridTemplateColumns}
      />,
    );

    // Check if all elements render correctly
    await screen.findByText('improvement.html');
    expect(await screen.findByText('Improvement')).toBeInTheDocument();
  });

  it('renders browser name when base and new apps are different', async () => {
    const { subtestsResult } = getTestData();
    const mockGridTemplateColumns = '1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr';
    renderWithRoute(
      <SubtestsRevisionRow
        result={subtestsResult[4]}
        gridTemplateColumns={mockGridTemplateColumns}
      />,
    );

    expect(document.body).toMatchSnapshot();
    expect(await screen.findByText(/firefox/i)).toBeInTheDocument();
    expect(await screen.findByText(/chrome/i)).toBeInTheDocument();
  });

  it('renders doesnt render browser name when base and new apps are firefox or fenix', async () => {
    const { subtestsResult } = getTestData();
    const mockGridTemplateColumns = '1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr';
    renderWithRoute(
      <SubtestsRevisionRow
        result={subtestsResult[0]}
        gridTemplateColumns={mockGridTemplateColumns}
      />,
    );

    expect(document.body).toMatchSnapshot();
    expect(screen.queryByText(/firefox/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/chrome/i)).not.toBeInTheDocument();
  });
});
