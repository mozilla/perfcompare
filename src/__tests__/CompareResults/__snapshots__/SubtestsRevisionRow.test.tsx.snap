// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtestsRevisionRow Component renders the component with correct data 1`] = `
<body>
  <div>
    <div
      class="revisionRow f15pv6go f1vh4jdk MuiBox-root css-1ykk23q"
      role="row"
    >
      <div
        class="subtests cell"
        role="cell"
      >
        dhtml.html
      </div>
      <div
        class="base-value cell"
        role="cell"
      >
         
        971.38
         
        ms
         
      </div>
      <div
        class="comparison-sign cell"
        role="cell"
      >
        &lt;
      </div>
      <div
        class="new-value cell"
        role="cell"
      >
         
        982.41
         
        ms
      </div>
      <div
        class="status cell"
        role="cell"
      >
        <div
          class="status-hint  MuiBox-root css-1dj4wrv"
        >
          -
        </div>
      </div>
      <div
        class="delta cell"
        role="cell"
      >
         
        1.14
         %
         
      </div>
      <div
        class="confidence cell"
        role="cell"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vsehry-MuiSvgIcon-root"
          data-testid="KeyboardArrowDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
        Low
      </div>
      <div
        class="total-runs cell"
        role="cell"
      >
        <span>
          <span
            title="Base runs"
          >
            B:
          </span>
          <strong>
            2
          </strong>
        </span>
        <span>
          <span
            title="New runs"
          >
            N:
          </span>
          <strong>
            2
          </strong>
        </span>
      </div>
      <div
        class="row-buttons cell"
      >
        <div
          class="graph"
          role="cell"
        >
          <div
            class="graph-link-button-container"
          >
            <a
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1itd531-MuiButtonBase-root-MuiIconButton-root"
              href="https://treeherder.mozilla.org/perfherder/graphs?highlightedRevisions=d775409d7c6a&highlightedRevisions=22f4cf67e8ad&series=mozilla-central%2C5c47b31c38f7214a07fad2e0d41fb901cdc18eae%2C1%2C1&timerange=604800"
              tabindex="0"
              target="_blank"
              title="open the evolution graph for this job in treeherder"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                data-testid="TimelineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
      <div
        class="expand-button cell"
        role="cell"
      >
        <button
          aria-controls=":r0:"
          aria-expanded="false"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-1itd531-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          title="expand this row"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
